src/main/java/com/firebolt/jdbc/connection/FireboltConnectionServiceSecret.java

```diff
@@ -0,0 +1,128 @@
+                                    FireboltEngineInformationSchemaService fireboltEngineService,
+                                    FireboltAccountIdService fireboltAccountIdService) throws SQLException {
+        super(url, connectionSettings, fireboltAuthenticationService, fireboltStatementService);
+        this.fireboltGatewayUrlService = fireboltGatewayUrlService;
+        this.fireboltAccountIdService = fireboltAccountIdService;
+        this.fireboltEngineService = fireboltEngineService;
+        connect();
+    }
+
+    @ExcludeFromJacocoGeneratedReport
+    FireboltConnectionServiceSecret(@NonNull String url, Properties connectionSettings) throws SQLException {
+        super(url, connectionSettings);
+        OkHttpClient httpClient = getHttpClient(loginProperties);
+        ObjectMapper objectMapper = FireboltObjectMapper.getInstance();
+        this.fireboltGatewayUrlService = new FireboltGatewayUrlService(createFireboltAccountRetriever(httpClient, objectMapper, "engineUrl", GatewayUrlResponse.class));
+        this.fireboltAccountIdService = new FireboltAccountIdService(createFireboltAccountRetriever(httpClient, objectMapper, "resolve", FireboltAccount.class));
+        this.fireboltEngineService = new FireboltEngineInformationSchemaService(this);
+        connect();
+    }
+
+    private <T> FireboltAccountRetriever<T> createFireboltAccountRetriever(OkHttpClient httpClient, ObjectMapper objectMapper, String path, Class<T> type) {
+        return new FireboltAccountRetriever<>(httpClient, objectMapper, this, loginProperties.getUserDrivers(), loginProperties.getUserClients(), loginProperties.getHost(), path, type);
+    }
+
+    @Override
+    protected void authenticate() throws SQLException {
+        String account = loginProperties.getAccount();
+        if (account == null) {
+            throw new FireboltException("Cannot connect: account is missing");
+        }
+        if (loginProperties.isSystemEngine() && loginProperties.getDatabase() == null) {
+            throw new FireboltException("The database with the name null could not be found", RESOURCE_NOT_FOUND);
+        }
+        String accessToken = getAccessToken(loginProperties).orElse(StringUtils.EMPTY);
+        sessionProperties = getSessionPropertiesForSystemEngine(accessToken, account);
+        assertDatabaseExisting(loginProperties.getDatabase());
+        if (!loginProperties.isSystemEngine()) {
+            sessionProperties = getSessionPropertiesForNonSystemEngine();
+        }
+    }
+
+    private FireboltProperties getSessionPropertiesForNonSystemEngine() throws SQLException {
+        sessionProperties = sessionProperties.toBuilder().engine(loginProperties.getEngine()).build();
+        Engine engine = fireboltEngineService.getEngine(loginProperties);
+        return loginProperties.toBuilder().host(engine.getEndpoint()).engine(engine.getName()).systemEngine(false).database(engine.getDatabase()).build();
```

Comment:
Here, we use engine.getDatabase as an engine database. Where do we use the database, provided in the connection string?

Reply:
Add an integration test, where you connect to an engine, and provide a database_name, that is invalid (no the same as the database it's actually connected to). This should fail

Reply:
done. `ConnectionTest.connectToWrongDb()`